src/rust/src/lib.rs

```diff
@@ -221,6 +221,20 @@ mod _rust {
                     openssl_mod.add("_legacy_provider_loaded", false)?;
                 }
             }
+            cfg_if::cfg_if! {
+                if #[cfg(CRYPTOGRAPHY_OPENSSL_320_OR_GREATER)] {
+                    use std::ptr;
+                    use std::cmp::max;
+
+                    let available = std::thread::available_parallelism().map_or(0, |v| v.get() as u64);
+                    // SAFETY: This sets a libctx provider limit, but we always use the same libctx by passing NULL.
+                    unsafe {
+                        let current = openssl_sys::OSSL_get_max_threads(ptr::null_mut());
+                        // Set the thread limit to the max of available parallelism or current limit.
+                        openssl_sys::OSSL_set_max_threads(ptr::null_mut(), max(available, current));
```

Comment:
Also, what happens if someone else has called `OSSL_set_max_threads` before us (e.g. CPython?)

Reply:
We'll call it again with the max of either the value they set or the max parallelism we detected (whatever is larger).

If the question is what happens if via a race condition we end up setting a max thread limit smaller than the current limit then `OSSL_set_max_threads` just updates a value in the `OSSL_LIB_CTX_THREADS` structure defined in `include/internal/thread.h`. The actual logic for the threadpool lives in `crypto/thread/internal.c`. If there's no active threads then changing it to a lower value is trivial, but if there are OpenSSL may have an issue? On initial review (without having tried to write a test to see if it actually does this):

* Set max threads to 10 at program start
* Spawn a thread (not in the OpenSSL threadpool) which calls argon2id with 40 lanes and enough iterations to take a bit of time.
* In main thread call `OSSL_set_max_threads` and set to 2
* Until the number of available threads drops to 2 or less any subsequent call to [_ossl_get_available_threads](https://github.com/openssl/openssl/blob/012353bdf21b98def920ac317b94c4a9ed501b79/crypto/thread/internal.c#L22) would return an underflowed value.

Reply:
Ok. I guess this is as good as can be done.